pkg/kubectl/cmd/drain.go

```diff
@@ -0,0 +1,159 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/labels"
+)
+
+type DrainOptions struct {
+	NodeName           string
+	Force              bool
+	GracePeriodSeconds int
+}
+
+const (
+	drain_long = `Drain nodes in preparation for maintenance.
+
+`
+	drain_example = `# Drain node "foo", even if there are unreplicated pods on it.
+$ kubectl drain foo --force
+
+# As above, but don't delete unreplicated pods, and use a grace period of 15 minutes.
+$ kubectl drain foo --grace-period=900
+`
+)
+
+func NewCmdDrain(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &DrainOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "drain NODE",
+		Short:   "Drain nodes in preparation for maintenance",
+		Long:    drain_long,
+		Example: drain_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(
+				RunDrain(f, out, cmd, args, options))
+		},
+	}
+	cmd.Flags().BoolVar(&options.Force, "force", false, "Delete unreplicated pods.")
+	cmd.Flags().IntVar(&options.GracePeriodSeconds, "grace-period", -1, "Period of time in seconds given to each pod to terminate gracefully. Ignored if negative.")
+	return cmd
+}
+
+func RunDrain(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string, options *DrainOptions) error {
+	if len(args) != 1 {
+		return cmdutil.UsageError(cmd, fmt.Sprintf("USAGE: %s [flags]", cmd.Use))
+	}
+	options.NodeName = args[0]
+
+	mapper, _ := f.Object()
+	c, err := f.Client()
+	if err != nil {
+		return err
+	}
+	if err = maybeLameNode(mapper, c, out, options); err != nil {
+		return err
+	}
+	pods, err := c.Pods(api.NamespaceAll).List(nil, fields.Set{"spec.nodeName": options.NodeName}.AsSelector())
+	if err != nil {
+		return err
+	}
+	if err = scanForReplicationControllers(c, out, options, pods); err != nil {
+		return err
+	}
+	if err = deletePods(mapper, c, out, options, pods); err != nil {
+		return err
+	}
+	cmdutil.PrintSuccess(mapper, false, out, "node", options.NodeName, "drained")
+	return nil
+}
+
+func maybeLameNode(mapper meta.RESTMapper, c *client.Client, out io.Writer, options *DrainOptions) error {
+	node, err := c.Nodes().Get(options.NodeName)
```

Comment:
Since you fetch the object in both commands, I would argue over using the resource builder.

Pros: 1) we can support both `kubectl drain foo` and `kubectl drain node/foo` syntax 2) you wouldn't need to add the kind explicitly in `PrintSuccess`.